luigi/contrib/spark.py

```diff
@@ -99,7 +83,247 @@ def __str__(self):
+                command += ['--jars', self.jars]
+        if self.py_files:
+            if isinstance(self.py_files, (list, tuple)):
+                command += ['--py-files', ','.join(self.py_files)]
+            elif isinstance(self.py_files, six.string_types):
+                command += ['--py-files', self.py_files]
+        if self.files:
+            if isinstance(self.files, (list, tuple)):
+                command += ['--files', ','.join(self.files)]
+            elif isinstance(self.files, six.string_types):
+                command += ['--files', self.files]
+        if self.archives:
+            if isinstance(self.archives, (list, tuple)):
+                command += ['--archives', ','.join(self.archives)]
+            elif isinstance(self.archives, six.string_types):
+                command += ['--archives', self.archives]
+        if self.conf:
+            conf = None
+            if isinstance(self.conf, six.string_types):
+                conf = dict(map(lambda i: i.split('='), self.conf.split('|')))
+            if isinstance(self.conf, dict):
+                conf = self.conf
+            if conf:
+                for prop, value in conf.items():
+                    command += ['--conf', '{0}="{1}"'.format(prop, value)]
+        if self.properties_file:
+            command += ['--properties-file', self.properties_file]
+        if self.driver_memory:
+            command += ['--driver-memory', self.driver_memory]
+        if self.driver_java_options:
+            command += ['--driver-java-options', self.driver_java_options]
+        if self.driver_library_path:
+            command += ['--driver-library-path', self.driver_library_path]
+        if self.driver_class_path:
+            command += ['--driver-class-path', self.driver_class_path]
+        if self.executor_memory:
+            command += ['--executor-memory', self.executor_memory]
+        if self.driver_cores:
+            command += ['--driver-cores', self.driver_cores]
+        if self.supervise:
+            command += ['--supervise']
+        if self.total_executor_cores:
+            command += ['--total-executor-cores', self.total_executor_cores]
+        if self.executor_cores:
+            command += ['--executor-cores', self.executor_cores]
+        if self.queue:
+            command += ['--queue', self.queue]
+        if self.num_executors:
+            command += ['--num-executors', self.num_executors]
+        return command
+
+    def app_command(self):
+        if not self.app:
+            raise NotImplementedError("subclass should define an app (.jar or .py file)")
+        return [self.app] + self.app_options()
+
+    def run(self):
+        args = map(str, self.spark_command() + self.app_command())
+        env = os.environ.copy()
+        temp_stderr = tempfile.TemporaryFile()
+        logger.info('Running: %s', repr(args))
+        proc = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=temp_stderr, env=env, close_fds=True)
```

Comment:
Added ``universal_newline=True`` to https://github.com/spotify/luigi/pull/837.

https://github.com/spotify/luigi/pull/837 adds a get_environment() that alters the env before passing it to ``Popen``.